src/audio/copier/copier.c

```diff
@@ -575,20 +555,69 @@ static struct comp_dev *copier_new(const struct comp_driver *drv,
 		default:
 			comp_cl_err(&comp_copier, "unsupported dma type %x",
 				    (uint32_t)node_id.f.dma_type);
-			goto error_cd;
+			goto error;
 		};
 
-		dev->direction_set = true;
+	dev->direction_set = true;
 	}
 
 	dev->direction = cd->direction;
 	dev->state = COMP_STATE_READY;
-	return dev;
+	return 0;
+error:
+	return -EINVAL;
+}
 
-error_cd:
+static struct comp_dev *copier_new(const struct comp_driver *drv,
+				   const struct comp_ipc_config *config,
+				   const void *spec)
+{
+	const struct ipc4_copier_module_cfg *copier = spec;
+	struct copier_data *cd;
+	struct comp_dev *dev;
+	struct processing_module *mod;
+	struct module_data *md;
+	int ret;
+
+	comp_cl_dbg(&comp_copier, "copier_new()");
+
+	mod = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM, sizeof(*mod));
+	if (!mod)
+		return NULL;
+
+	dev = comp_alloc(drv, sizeof(*dev));
+	if (!dev)
+		goto error_dev;
+
+	dev->ipc_config = *config;
+
+	cd = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM, sizeof(*cd));
+	if (!cd)
+		goto error_cd;
+
+	comp_set_drvdata(dev, mod);
```

Comment:
@btian1 if you do this here, you have to change all the comp_get_drv_data() in the code to use 
struct copier_data *cd = module_get_private_data(mod);

Reply:
I know, will modify all of them later.